modules/stochastic_tools/include/reporters/SnapshotContainerBase.h

```diff
@@ -0,0 +1,62 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "GeneralReporter.h"
+#include "libmesh/petsc_vector.h"
+#include "libmesh/petsc_matrix.h"
+#include "NonlinearSystemBase.h"
+
+/**
+ * Base class for storing and managing numerical data like solutions, residuals, and Jacobians.
+ * The vectors are kept distributed with respect to the communicator of the application.
+ * The whole snapshot vector is stored. The saving frequency can be defined using the `execute_on`
+ * parameter.
+ */
+class SnapshotContainerBase : public GeneralReporter
+{
+public:
+  static InputParameters validParams();
+  SnapshotContainerBase(const InputParameters & parameters);
+  virtual void initialSetup() override;
+  virtual void initialize() override {}
+  virtual void execute() override;
+  virtual void finalize() override {}
+
+  /**
+   * Return the whole snapshot container
+   * @return const std::vector<std::unique_ptr<NumericVector<Number>>>&
+   */
+  const std::vector<std::unique_ptr<NumericVector<Number>>> & getContainer()
+  {
+    return _accumulated_data;
+  }
+
+  /**
+   * Return one of the stored solution vectors
+   * @param local_i The index of the locally stored solution vector
+   */
+  const std::unique_ptr<NumericVector<Number>> & getSnapshot(unsigned int local_i) const;
+
+protected:
+  /**
+   * Clone the current snapshot vector.
+   * @return std::unique_ptr<NumericVector<Number>>
+   */
+  virtual std::unique_ptr<NumericVector<Number>> cloneSnapshot() = 0;
```

Comment:
When will the types change? You typically need a clone method when you have derived classes of different underlying types. I don't quite follow how that fits here

Reply:
Might not be following. I need to get the snapshots of the residual vector and the jacobian matrix(row concatenated). The new derived classes (ex:ResidualContainer) would only need to override the cloneSnapshot method to be able to get the snapshots and prepare for transfering.

Reply:
Yeah I was not aware of any standardized way of using the `clone` method. I would not transition to `getSnapshot` either. Hmmm, how about `collect/draw/pick/gatherSnapshot`? Gather might be another collision with MPI routines.